### DISCUSSION TITLE
Where the users are, August 2021

### DISCUSSION THREAD

**Discussion by rossabaker** [http4s/http4s#5245] [Show and tell]
Based on Maven Central stats for August 2021.

# http4s / Scala pairs

||_2.11|_2.12|_2.13|_3|Total|
|----:|----:|----:|----:|----:|----:|
|0.18.x|0%|3%| | |3%|
|0.20.x|1%|3%| | |3%|
|0.21.x|0%|24%|43%| |67%|
|0.22.x| |5%|3%|0%|8%|
|0.23.x| |1%|6%|1%|9%|
|1.0.x| |1%|5%|0%|6%|
|Total|3%|38%|58%|2%| |

9% of our core downloads are getting no security patches (pre-0.21).  77% of our core downloads are not getting regular bugfixes (pre-0.22).

Scala 2.13 is now a clear majority, with our absolute 2.12 numbers in decline for the first time.  Scala 3 uptake continues to lag.

Cats-Effect 2 on Scala 3 is less than 1% of our users.

# http4s usage by Scala compiler

Read down: columns sum to 100% (minus weird versions):

| |_2.11|_2.12|_2.13|_3|
|----:|----:|----:|----:|----:|
|0.18.x|14%|8%| | |
|0.20.x|31%|7%| | |
|0.21.x|0%|64%|74%| |
|0.22.x| |12%|6%|7%|
|0.23.x| |4%|11%|70%|
|1.0.x| |2%|9%|24%|

94% of Scala 3 users are already on Cats-Effect 3.

# Scala compiler by http4s version

Read across: rows sum to 100% (minus weird versions):

| |_2.11|_2.12|_2.13|_3|
|----:|----:|----:|----:|----:|
|0.18.x|12%|88%| | |
|0.20.x|27%|73%| | |
|0.21.x|0%|36%|64%| |
|0.22.x| |56%|42%|1%|
|0.23.x| |16%|71%|13%|
|1.0.x| |13%|81%|6%|

Scala 2.13 is the sweet spot for everyone but http4s-0.22 users.  Are people stopping at 0.22 because of Spark?

**Comment by rossabaker** [comment]
This calls into question the viability of 0.22.

* Lots of people are stuck on 0.21.  @zmccoy and I are looking at docs and scalafixes to assist mass migration at $WORK.  But ours will be straight to 0.23.
* Scala 3 forces people past 0.21, but that has not enticed many people yet.
* Cats-Effect 3 forces people past 0.21, but those people must skip 0.22.

0.22 was intended to be the waystation for Scala 3, but Cats-Effect 3 will be dominant before Scala 3 becomes significant.  We should probably scale our 0.22 efforts back to bugfixes and community-supported backports.

**Reply by armanbilge** [reply]
Not sure if it's entirely relevant to this discussion re 0.22, but based on some of my recent PRs I think we can dramatically tighten the diff between 0.23 and 1.0. I think the JS port is the only major difference and lots of upstream cross-building in fs2.io.file and http4s-crypto (and ex-communicating dom) has vanished many of the pain points.

**Reply by RafalSumislawski** [reply]
For me the question is if we should keep providing bugfixes for 0.21 where the majority of users is, or if we should "motivate" that crowd to migrate to 0.22 by not providing these fixes... Other than that there's not much incentive to migrate to 0.22.

**Reply by rossabaker** [reply]
Keeping the active branches as similar as possible helps our forward merges, so I agree with Arman's strategy.

To Rafał's point, we've got a few tiers of support.  We have a [security.md](https://github.com/http4s/http4s/blob/main/SECURITY.md), and separately, the [versions page](https://http4s.org/versions).  0.21 is in an awkward stage where it's "EOL" but got three releases this month.

|      | security.md | /versions   | Security patches | Bincompat dependencies | Bincompat  bugfixes | Bincompat  features | Breaking  changes |
|------|-------------|-------------|------------------|------------------------|-----------------|---------------------|-------------------|
| 0.20 | 🚫           | EOL         | 🚫                | 🚫                      | 🚫               | 🚫                   | 🚫                 |
| 0.21 | ✅           | EOL         | ✅                | 🚫                       | 🚫               | 🚫                   | 🚫                 |
| 0.22 | ✅           | Stable      | ✅                | ✅                      | ✅               | ✅mostly    | 🚫                 |
| 0.23 | ✅           | Stable      | ✅                | ✅                      | ✅               | ✅                   | 🚫                 |
| 1.0  | ✅           | Development | ✅                | ✅                      | ✅               | ✅                   | ✅                 |

I'd like there to not be two kinds of "EOL".  Just one place to look.  I'd like to bury 0.21, but my work is still mostly on it.  Maybe there's something missing between "EOL" and "Stable" that would be appropriate for 0.22 (or even 0.23): bugfixes without features?  I want to spend less time merging, but also give people nice things while 1.0's more radical ideas mature.  I think we need to revisit the above table and then refine the public promises.

**Reply by armanbilge** [reply]
> Keeping the active branches as similar as possible helps our forward merges, so I agree with Arman's strategy.

"Active" being the key word here. The more we do to unify 0.23 and 1.0 the further they get from 0.22. But if 0.22 is no longer "active", that opens the door to be more aggressive about bringing 0.23 and 1.0 back together.

> Maybe there's something missing between "EOL" and "Stable" that would be appropriate for 0.22 (or even 0.23): bugfixes without features?

Maybe the missing stage is "maintained"? Something like:
* development (bin-breaking features) 
* stable (bin-compat features)
* maintained (bug fixes)
* EOL (archived)

**Reply by rossabaker** [reply]
This is what feels right to me, but it's five tiers, and I don't want five.  Help me refine this:

|      | Characterization | EOL        | Security patches | Bincompat dependencies | Bincompat  bugs | Bincompat  features | Optimizations | Breaking  changes |
|------|------------------|------------|------------------|------------------------|-----------------|---------------------|---------------|-------------------|
| 0.20 | EOL              | 2021-02-09 | 🚫                | 🚫                      | 🚫               | 🚫                   | 🚫             | 🚫                 |
| 0.21 | Deprecated       | 2022-07-20 | ✅                | 🚫                      | 🚫               | 🚫                   | 🚫             | 🚫                 |
| 0.22 | Legacy           | 2022-07-30 | ✅                | ✅                      | ✅               | 🚫                   | 🚫             | 🚫                 |
| 0.23 | Production       | TBD        | ✅                | ✅                      | ✅               | ✅                   | ✅             | 🚫                 |
| 1.0  | Development     | TBD        | ✅                | ✅                      | ✅               | ✅                   | ✅             | ✅                 |

* EOL: No further releases. You're on your own.  No sooner than 12 months after a subsequent production release.
* Deprecated: Move off ASAP. Only security-related patches are planned.
* Legacy: We'll fix bugs, but not aggressively backport features.
* Production: Where most projects should aspire to be.
* Development: No compatibility guarantees

"Deprecated" and "Legacy" are confusing, but I want to give users more than two months to get off 0.21, and it's shitty to abandon 0.22 users without routine maintenance just two months after its release.  How can we simplify this and not be jerks?

**Reply by bplommer** [reply]
How about "obsolete" / "EOL" instead of "EOL" / "deprecated"?

**Reply by rossabaker** [reply]
I think of "EOL" as truly dead.  Some people use "sunset" for things that are about dead.

Can someone convince me that there are less than five tiers?  0.21 and 0.22 are similar in that there is something better and also stable, but they're different in that we were selling people a shiny new 0.22 in July.  

We're only in this position because of the exceptional circumstances of Scala 3 and Cats-Effect 3 looming about the same time and needing to get on both without being sure which would get traction first.  Maybe one of these tiers is exceptional.  I certainly don't want to repeat this position.

**Reply by hamnis** [reply]
A colleage of mine just did upgrade to scala 3 through 0.22 since doing that + CE 3 at the same time was too big a task.

**Reply by rossabaker** [reply]
If 0.22 gets bugfixes, but not the shiniest new features, is it just as good?  I think that would make 0.21 -> 0.22 easier, but 0.22 -> later harder.

**Reply by bplommer** [reply]
I found 0.22 valuable for the first couple of services I updated, not because of the Scala 3 thing but because 0.23 is largely source-compatible with it modulo cats-effect and so I didn't have to update to CE3 at the same time as dealing with breaking changes to http4s.

> If 0.22 gets bugfixes, but not the shiniest new features, is it just as good?

I think so, yes

> I think that would make 0.21 -> 0.22 easier, but 0.22 -> later harder.

Both of these only apply if the shiny new features are source-breaking, right? @armanbilge convinced me in #5229 that they shouldn't be, though I'd personally be more than happy for you to overrule that.

**Reply by armanbilge** [reply]
> 0.22 -> later harder

With hindsight, I am increasingly apologetic that I eviscerated 1.0 with the JS port. The number of breaking changes it needed I can list on my fingers (and mostly for minor things) and the file splitting was driven by lack of crossed fs2.io and crypto for JS when I started, which has also been addressed. The best time to correct this situation is before any other significant 1.0 development occurs so that we can "recut" a 1.0 that looks much more like 0.23.

We can fix all this, but then 0.22 will be left fairly far behind. (The problem is the lack of the new fs2.io files APIs in the CE2 series.)

**Reply by bplommer** [reply]
> I think that would make 0.21 -> 0.22 easier, but 0.22 -> later harder.

Wait, are we talking about ease of user upgrades or ease of merges?

**Reply by rossabaker** [reply]
> Both of these only apply if the shiny new features are source-breaking, right?

Yes, true.

When we make source-breaking changes on patch releases, it causes user fatigue, and it's dangerous when we add friction to applying a security patch.  (This week's, for instance, should be a simple build.sbt bump for most people.)  I wouldn't say there's an absolute rule against it like there is on binary, but they need to carry a lot of weight.

**Reply by armanbilge** [reply]
That's a good question, I assumed we were talking merges.

**Reply by rossabaker** [reply]
Note that we just got a run-of-the-mill bugfix to 0.21.  I don't mind releasing that, because our automation is good, and our process is currently confusing as fuck.

**Comment by rossabaker** [comment]
The elephant in the room is Scala 2.12 support. 

It's established on 0.22 and 0.23, and there's little reason to drop it.

Absolute demand for 1.0 on Scala 2.12 will likely rise: all those 0.21 users eventually have to go somewhere, and those are the conservatives.  And generous crossbuilds are something that those of us on large projects appreciate.  But even on http4s-0.23, a production release, Scala 2.12 usage is down to 16%.  And the Spark Line is about to progress forward to 2.13.  It's really not causing any friction for us, but I'm getting less willing to piss off upstream maintainers in its defense.

**Reply by armanbilge** [reply]
Dropping 2.12 would make the way for some of @bplommer's work relying on singleton types.

**Reply by bplommer** [reply]
It's not just singleton types - 2.13 has generally better type inference than 2.12. I'm working on a PR to make `Message` and its subtypes covariant in `F`, which almost completely eliminates the need to specify effect types when creating `Request` and `Response` instances in Scala 2.13 but doesn't work so well in Scala 2.12.

**Reply by rossabaker** [reply]
Okay, that would be the most compelling case yet.  Let's see where it goes.

**Comment by listba** [comment]
I wanted to chime in here with my personal experience. The main reason I haven't upgraded past `0.21.x` is that I honestly didn't know `0.22.x` and `0.23.x` were even considered stable until now. One of the driving reasons for that is the old doc site of `0.21.x` which i have bookmarked and several google searches still lead directly too (such as `http4s client` or `http4s server`) still list `0.21.x` as the stable version, and doesn't even show `0.22.x` or `0.23.x`
![image](https://user-images.githubusercontent.com/777962/142479369-c8c45755-1f9e-41df-b51d-c0227370fba4.png)

The other reason for us personally is we rely on the `rho_routes` swagger generation and there are only Milestone/ RC's for `0.22.x` and `0.23.x`

**Reply by rossabaker** [reply]
Thanks, that's great feedback.  There's no dynamic call from a central source, so we have to republish all the old versions.  I would have thought that would have been fixed with one of the CVEs, but I guess not.

**Reply by listba** [reply]
Looks like it should have been fixed by this mr but I don't know how your publishing works: https://github.com/http4s/http4s/pull/5481
